doc/netns.8
.TH NETNS 8
.SH NAME
netns \- sets up and configures a new network namespace
.SH SYNOPSIS
.B netns [-hv] start|getpid 
.I NAMESPACE
.SH DESCRIPTION
.B netns
is usually used to initialize network (and UTS) namespaces. It executes two
scripts; 'assign' is executed in the original namespace while 'start' is
executed in the newly created namespace.  
.SH OPTIONS
.IP -v
Enable verbose output
.IP start
Start the namespace specified with
.I NAMESPACE
.IP getpid
Prints the namespace pid of a running namespace
.IP -h
Outputs usage information
.SH FILES
.I /etc/netns/NAMESPACE
.RS
The directory containing all configuration of a single namespace
.RE
.I /etc/netns/NAMESPACE/assign
.RS
Contains the host system part of the setup procedures
.RE
.I /etc/netns/NAMESPACE/start
.RS
Contains the child namespace part of the setup procedures
.RE
.SH ENVIRONMENT

.IP NETNS_NAME
The ASCII name of the namespace
.IP NETNS_PID
The process ID of the newly spawned namespace
.SH EXAMPLE
The following example demonstrates the use of 
.BR netns
to spawn a fairly basic network namespace. 
.SS Assign script
\&
.nf
#! /bin/sh
ip link add name veth0 type veth peer name veth.netns
ip link set veth0 netns $NETNS_PID
ip addr add 172.22.22.1/30 dev veth.netns scope link
ip link set veth.netns up
.fi
.SS Start script
\&
.nf
#! /bin/sh
hostname foo-netns
ip link set lo up
ip addr add 172.22.22.2/30 dev veth0
ip link set veth0 up
screen -d -m
.fi

After starting the netns there should be a detached screen session you can
attach to for further configuration. A lightweight, init-like system that
could be used for starting processes is
.BR runsvdir (2)
included in runit.
.SH BUGS
TBD. The software itself should be more or less free of bugs as it is fairly
simple. Most problems encountered are caused by using older kernel versions.
Almost all bugs with regard to network namespaces should be fixed in 2.6.35.
.SH AUTHOR
netns was written by Daniel Lamparter <equinox at diac24 dot net>

This man page was written by Benjamin Kiessling <benjaminkiessling at bttec dot org>
.SH "SEE ALSO"
.BR clone (2)



src/netns.c
#define _GNU_SOURCE
#include "acconf.h"

#include <stdio.h>
#include <stdlib.h>
#include <sched.h>
#include <string.h>
#include <errno.h>
#include <libgen.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/poll.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>

#ifdef HAVE_SCHED_H
#include <sched.h>
#endif
#ifdef HAVE_LINUX_UNISTD_H
#include <linux/unistd.h>
#endif

#ifndef PATH_CFG
#error define PATH_CFG
#endif
#ifndef PATH_STATE
#error define PATH_STATE
#endif

#include "netns.h"

static int verbose = 0;

static int usage(int rv)
{
	FILE *fd = rv ? stderr : stdout;
	fprintf(fd, "\
Usage:	netns -h\n\
	netns [-v] <OPERATION> <NAMESPACE NAME>\n\n\
where OPERATION can be:\n\
	start		start the namespace\n\
	getpid		print the namespace pid if it is running\n\
\n\
");
	return rv;
}

static int launch_cmd(const char *exe, char * const *args)
{
	pid_t pid;
	int fd;
	int status;

	pid = fork();
	switch (pid) {
	case -1:
		perror("fork");
		return -1;
	case 0:
		for (fd = 3; fd < 10; fd++)
			close(fd);
		if (verbose >= 1)
			fprintf(stderr, "launching %s\n", exe);
		execv(exe, args);
		fprintf(stderr, "%s: %s\n", exe, strerror(errno));
		exit(1);
	default:
		if (waitpid(pid, &status, 0) < 0) {
			perror("waitpid");
			return 1;
		}
		return WIFEXITED(status) ? WEXITSTATUS(status) : -1;
	}
}

static char pidpath[PATH_MAX];
static pid_t mainpid;

static void removepid()
{
	if (getpid() == mainpid)
		unlink(pidpath);
}

static int writepid(const char *ns)
{
	char pid[32];
	int pidfd;
	struct flock lck;

	snprintf(pidpath, sizeof(pidpath), PATH_STATE"/%s/pid", ns);
	pidfd = open(pidpath, O_RDWR | O_NONBLOCK | O_TRUNC | O_CREAT, 0644);
	if (pidfd < 0) {
		fprintf(stderr, "%s: %s\n", pidpath, strerror(errno));
		return 1;
	}

	mainpid = getpid();
	snprintf(pid, sizeof(pid), "%ld\n", (long)mainpid);
	safe_write(pidfd, pid, strlen(pid));

	lck.l_type = F_WRLCK;
	lck.l_whence = SEEK_SET;
	lck.l_start = 0;
	lck.l_len = strlen(pid);
	lck.l_pid = 0;
	if (fcntl(pidfd, F_SETLK, &lck) < 0) {
		fprintf(stderr, "fcntl(%s, F_SETLK): %s\n", pidpath,
				strerror(errno));
		close(pidfd);
		unlink(pidpath);
		return 1;
	}
	atexit(removepid);
	return 0;
}

static int do_start_child(const char *ns, int notifier, int waiter)
{
	int ok = 0, rv, ctlsock;
	struct pollfd p;

	char exe[256], argv0[256], *args[] = { argv0, NULL };

	int sig;
	sigset_t set;

	/* write pid first to catch errors */

	rv = writepid(ns);
	if (rv) {
		safe_write(notifier, &rv, sizeof(rv));
		return 1;
	}
	ctlsock = ctl_open(ns);

	if (unshare(CLONE_NEWNET | CLONE_NEWUTS) == -1) {
		perror("unshare");
		rv = 1;
		safe_write(notifier, &rv, sizeof(rv));
		return 1;
	}
	if (verbose >= 1)
		fprintf(stderr, "%s: pid %ld\n", ns, (long)getpid());

	safe_write(notifier, &ok, sizeof(ok));

	/* master now runs /etc/netns/(ns)/assign */

	p.fd = waiter;
	p.events = POLLIN;
	if (poll(&p, 1, -1) <= 0) {
		perror("poll");
		return 1;
	}
	safe_read(waiter, &rv, sizeof(rv));
	if (rv)
		return 1;

	close(waiter);

	setsid();

	snprintf(argv0, sizeof(argv0), "netns: %s", ns);
	snprintf(exe, sizeof(exe), PATH_CFG "/%s/start", ns);
	rv = launch_cmd(exe, args);
	if (rv) {
		fprintf(stderr, "%s exited with status %d\n", exe, rv);
		safe_write(notifier, &rv, sizeof(rv));
		return 1;
	}

	if (verbose >= 1)
		fprintf(stderr, "%s: up and running\n", ns);

	safe_write(notifier, &ok, sizeof(ok));
	close(notifier);

	close(0);
	close(1);
	close(2);

	ctl_run(ctlsock);

	/* hmm. control socket bailed, just wait forever... */
	sigemptyset(&set);
	sigaddset(&set, SIGCHLD);
	while (1)
		sigwait(&set, &sig);
}

static int do_start_parent(const char *ns, int waiter, int notifier, int pid)
{
	int status = -1, ok = 0;
	char spid[10];
	struct pollfd p;

	char exe[256], argv0[256], *args[] = { argv0, NULL };

	snprintf(spid, sizeof(spid), "%d", pid);
	setenv("NETNS_PID", spid, 1);
	setenv("NETNS_NAME", ns, 1);

	/* child does unshare() */

	p.fd = waiter;
	p.events = POLLIN;
	if (poll(&p, 1, -1) <= 0) {
		perror("poll");
		return 1;
	}
	safe_read(waiter, &status, sizeof(status));
	if (status) {
		fprintf(stderr, "child flagged an error, aborting.\n");
		return 1;
	}

	snprintf(argv0, sizeof(argv0), "netns-assign: %s", ns);
	snprintf(exe, sizeof(exe), PATH_CFG "/%s/assign", ns);
	status = launch_cmd(exe, args);
	if (status) {
		fprintf(stderr, "%s exited with status %d\n", exe, status);
		safe_write(notifier, &status, sizeof(status));
		return 1;
	}

	safe_write(notifier, &ok, sizeof(ok));

	if (poll(&p, 1, -1) <= 0) {
		perror("poll");
		return 1;
	}
	safe_read(waiter, &status, sizeof(status));
	if (status) {
		fprintf(stderr, "child flagged an error, aborting.\n");
		return 1;
	}
	return 0;
}

static int do_start(const char *ns)
{
	pid_t pid;
	int sync1[2], sync2[2];

	if (pipe(sync1) < 0 || pipe(sync2) < 0) {
		perror("pipe");
		return 1;
	}

	pid = fork();
	switch (pid) {
	case -1:
		perror("fork");
		return 1;
	case 0:
		close(sync1[0]);
		close(sync2[1]);
		return do_start_child(ns, sync1[1], sync2[0]);
	default:
		close(sync1[1]);
		close(sync2[0]);
		return do_start_parent(ns, sync1[0], sync2[1], pid);
	}
}

static int do_getpid(const char *ns)
{
	int pidfd;
	struct flock lck;

	snprintf(pidpath, sizeof(pidpath), PATH_STATE"/%s/pid", ns);
	pidfd = open(pidpath, O_RDWR | O_NONBLOCK);
	if (pidfd < 0) {
		fprintf(stderr, "%s: %s\n", pidpath, strerror(errno));
		return 1;
	}

	lck.l_type = F_WRLCK;
	lck.l_whence = SEEK_SET;
	lck.l_start = 0;
	lck.l_len = 1;
	lck.l_pid = 0;
	if (fcntl(pidfd, F_GETLK, &lck) < 0) {
		fprintf(stderr, "fcntl(%s, F_GETLK): %s\n", pidpath,
				strerror(errno));
		return 1;
	}

	if (lck.l_type == F_WRLCK) {
		printf("%d", lck.l_pid);
		return 0;
	}
	fprintf(stderr, "%s: not running (pid file not locked)\n", ns);
	return 2;
}

int main(int argc, char * const argv[])
{
	int c;
	const char *cmd, *ns;
	char nspath[PATH_MAX];
	struct stat ns_stat;

	do {
		c = getopt(argc, argv, "hv");
		switch (c) {
		case -1:
			break;
		case 'h':
			return usage(0);
		case 'v':
			verbose++;
			break;
		default:
			return usage(1);
		}
	} while (c != -1);

	if (optind != argc - 2) {
		fprintf(stderr, "invalid number of arguments.\n");
		return 1;
	}
	cmd = argv[optind];
	ns = argv[optind + 1];

	if (strchr(ns, '/'))
		fprintf(stderr, "%s: namespace names containing slashes "
				"are not recommended.\n", ns);
	if (ns[0] == '.')
		fprintf(stderr, "%s: namespace names starting with dots "
				"are not recommended.\n", ns);

	snprintf(nspath, sizeof(nspath), PATH_CFG "/%s", ns);
	if (stat(nspath, &ns_stat)) {
		fprintf(stderr, "%s: %s\n", nspath, strerror(errno));
		return 2;
	}
	if (!S_ISDIR(ns_stat.st_mode)) {
		fprintf(stderr, "%s: not a directory\n", nspath);
		return 2;
	}

	if (verbose >= 1)
		fprintf(stderr, "%s: 0%o %d:%d\n", nspath, ns_stat.st_mode,
				ns_stat.st_uid, ns_stat.st_gid);

	if (!strcmp(cmd, "start")) {
		return do_start(ns);
	} else if (!strcmp(cmd, "getpid")) {
		return do_getpid(ns);
	} else {
		fprintf(stderr, "unknown command '%s'\n", cmd);
		return 1;
	}
}

